_posts/2023-02-28-kani-internship-projects-2022-stubbing.md

```diff
@@ -0,0 +1,361 @@
+In addition, Kani must be called with options `--enable-unstable --enable-stubbing --harness <harness_name>`.
+Note that `--harness` is needed because the stubbing feature is limited to a single harness at the moment.
+
+### An example: stubbing `rand::random`
+
+Let's see a simple example where we use the [`rand::random`](https://docs.rs/rand/latest/rand/fn.random.html) function
+to generate an encryption key.
+
+```rust
+#[cfg(kani)]
+#[kani::proof]
+fn encrypt_then_decrypt_is_identity() {
+    let data: u32 = kani::any();
+    let encryption_key: u32 = rand::random();
+    let encrypted_data = data ^ encryption_key;
+    let decrypted_data = encrypted_data ^ encryption_key;
+    assert_eq!(data, decrypted_data);
+}
+
+```
+
+At present, Kani fails to verify this example because its [FFI](https://doc.rust-lang.org/nomicon/ffi.html) support requires improvements (see [issue #1781](https://github.com/model-checking/kani/issues/1781) for more details).
+In other words, the code under verification contains unsupported features.
+
+However, the stubbing feature allows us to work around this limitation[^footnote-limitation] as follows:
+
+```rust
+#[cfg(kani)]
+fn stub_random<T: kani::Arbitrary>() -> T {
+    kani::any()
+}
+
+#[cfg(kani)]
+#[kani::proof]
+#[kani::stub(rand::random, stub_random)]
+fn encrypt_then_decrypt_is_identity() {
+    let data: u32 = kani::any();
+    let encryption_key: u32 = rand::random();
+    let encrypted_data = data ^ encryption_key;
+    let decrypted_data = encrypted_data ^ encryption_key;
+    assert_eq!(data, decrypted_data);
+}
+```
+
+Here, the `#[kani::stub(rand::random, stub_random)]` attribute indicates to Kani that it should replace `rand::random` with the stub `stub_random`.
+Note that this is a sound assumption to do: `rand::random` is expected to return any `u32` value, just like `kani::any`.
+
+Now, let's run it through Kani:
+
+```bash
+cargo kani --enable-unstable --enable-stubbing --harness encrypt_then_decrypt_is_identity
+```
+
+The verification result is composed of a single check: the assertion corresponding to `assert_eq!(data, decrypted_data)`.
+
+```
+RESULTS:
+Check 1: encrypt_then_decrypt_is_identity.assertion.1
+         - Status: SUCCESS
+         - Description: "assertion failed: data == decrypted_data"
+         - Location: src/main.rs:18:5 in function encrypt_then_decrypt_is_identity
+
+
+SUMMARY:
+ ** 0 of 1 failed
+
+VERIFICATION:- SUCCESSFUL
+```
+
+Kani verifies the assertion successfully, avoiding any issues that appear if we attempt to verify the code without stubbing.
+
+### Another example: stubbing `factorial`
+
+Let's see another example where we compute the [factorial](https://en.wikipedia.org/wiki/Factorial) of a number `n`.
+Note that the `factorial` function uses the method `checked_mul` to prevent overflow errors.
+We're interested in verifying the correctness of the `factorial` calculation.
```

Comment:
I think this example is a little weak for two reasons:

1. The stub implementation is clearly more efficient than the actual implementation, which raises the question: "Why not just use the stub as the actual implementation?"
2. If the point is to prove the correctness of the `factorial` implementation, then using a stub is self-defeating here (since we're just proving something about the stub instead).

One possibility is to verify something that involves factorial instead of factorial itself, such as proving something about binomial coefficients; for example, `(n choose k) == (n choose (n - k))`. It'd still leave the first problem, but I think that's less worrisome (this is clearly an example and not meant to be totally realistic).

Reply:
That's an excellent suggestion, thanks @aaronbembenek ! This will address part of @jimgrundy 's feedback as well.